### ISSUE TITLE
RPC: `unreachable!()` is not unreachable

### ISSUE THREAD

**Issue by sisou** [nimiq/core-rs-albatross#751]
**Update**: The specific conditions under which this issue occurs are described [in this comment](https://github.com/nimiq/core-rs-albatross/issues/751#issuecomment-1085673344).

---

Calling the `get_validator_by_address` RPC method, which internally calls `StakingContract::get_validator_stakers(...)`, panics at unreachable code:
https://github.com/nimiq/core-rs-albatross/blob/7c2a7c25e26c3d4fcc5d123db3aa57ee2d09f750/primitives/account/src/staking_contract/mod.rs#L195-L202

~~I remember that this did not panic at the last release candidate, so this might have something to do with the switch to MDBX and cursors working differently,~~ or have a totally different cause.

<details>
<summary>Stack trace</summary>

```
2022-03-31T08:41:14.038062Z ERROR panic                | thread 'tokio-runtime-worker' panicked at 'internal error: entered unreachable code': primitives/account/src/staking_contract/mod.rs:199
   0: log_panics::init::{{closure}}
   1: std::panicking::rust_panic_with_hook
             at /rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/std/src/panicking.rs:610:17
   2: std::panicking::begin_panic_handler::{{closure}}
             at /rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/std/src/panicking.rs:500:13
   3: std::sys_common::backtrace::__rust_end_short_backtrace
             at /rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/std/src/sys_common/backtrace.rs:139:18
   4: rust_begin_unwind
             at /rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/std/src/panicking.rs:498:5
   5: core::panicking::panic_fmt
             at /rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/core/src/panicking.rs:116:14
   6: core::panicking::panic
             at /rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/core/src/panicking.rs:48:5
   7: nimiq_account::staking_contract::StakingContract::get_validator_stakers
   8: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
   9: <nimiq_rpc_server::dispatchers::blockchain::BlockchainDispatcher as nimiq_jsonrpc_server::Dispatcher>::dispatch::{{closure}}
  10: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
  11: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
  12: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
  13: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
  14: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
  15: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
  16: <futures_util::future::try_maybe_done::TryMaybeDone<Fut> as core::future::future::Future>::poll
  17: <futures_util::future::try_join::TryJoin<Fut1,Fut2> as core::future::future::Future>::poll
  18: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
  19: <futures_util::future::try_future::try_flatten::TryFlatten<Fut,<Fut as futures_core::future::TryFuture>::Ok> as core::future::future::Future>::poll
  20: <futures_util::future::future::map::Map<Fut,F> as core::future::future::Future>::poll
  21: tokio::runtime::task::core::CoreStage<T>::poll
  22: std::panic::catch_unwind
  23: tokio::runtime::task::harness::poll_future
  24: tokio::runtime::task::harness::Harness<T,S>::poll
  25: std::thread::local::LocalKey<T>::with
  26: tokio::runtime::thread_pool::worker::Context::run_task
  27: tokio::runtime::thread_pool::worker::Context::run
  28: tokio::macros::scoped_tls::ScopedKey<T>::set
  29: tokio::runtime::thread_pool::worker::run
  30: <tokio::runtime::blocking::task::BlockingTask<T> as core::future::future::Future>::poll
  31: tokio::runtime::task::harness::Harness<T,S>::poll
  32: tokio::runtime::blocking::pool::Inner::run
  33: std::sys_common::backtrace::__rust_begin_short_backtrace
  34: core::ops::function::FnOnce::call_once{{vtable.shim}}
  35: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/alloc/src/boxed.rs:1854:9
      <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/alloc/src/boxed.rs:1854:9
      std::sys::unix::thread::Thread::new::thread_start
             at /rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/std/src/sys/unix/thread.rs:108:17
  36: start_thread
  37: __GI___clone3
```
</details>

**Comment by sisou** [comment]
Further investigation shows that when `validator.num_stakers == 0`, the validator address itself is returned in the accounts tree chunk (account type `Account::StakingValidator`). If the validator has a staker, only that staker is returned and `unreachable!()` is _not_ triggered.

**Comment by sisou** [comment]
I'm unable to reproduce it in these adapted tests, but it _is_ happening with my client in the devnet.
https://github.com/nimiq/core-rs-albatross/compare/sisou/validator-stakers?expand=1

(The CI failed because of formatting, not because of test failure.)

**Comment by sisou** [comment]
I managed to write a failing test that triggers it. You can see it [in this commit](https://github.com/nimiq/core-rs-albatross/commit/808c8f45225e18b4a14bcd625e7d542873dfcb82) on the branch [`sisou/validator-stakers`](https://github.com/nimiq/core-rs-albatross/tree/sisou/validator-stakers).

[The very last `assert_eq` in the test triggers the panic.](https://github.com/nimiq/core-rs-albatross/commit/808c8f45225e18b4a14bcd625e7d542873dfcb82#diff-88a7f2d41491c288638c5acdca790b065b399d5bed5c88efa426ed694850f518R384-R387) If that getter and check is removed, the test passes.

The "unreachable" code is **only triggered** when getting the stakers of a validator that has no stakers, **while** at least another validator with at least one staker **also exists**. That's why it gets triggered in the devnet, where validators with stakers exist in the genesis block, when I created and then queried my own validator.

Run the test isolated locally with
```
cargo test --package nimiq-account -- can_get_validator_stakers
```
---

A simple workaround would be to return early from the `get_validator_stakers` method, when it is found that `validator.num_stakers == 0`. But that would not solve the root issue with the database cursors, which might also have an effect elsewhere in the code.
https://github.com/nimiq/core-rs-albatross/blob/7c2a7c25e26c3d4fcc5d123db3aa57ee2d09f750/primitives/account/src/staking_contract/mod.rs#L185-L187

**Comment by viquezclaudio** [comment]
I ran the unitest included in the  _validator-stakers_ branch at various points of time, including the commit before removing LMDB and also the three devnet RC releases and in all those cases the test fails. 
So it seems the issue is more related towards stacking and is not a problem with the database implementation.

**Comment by styppo** [comment]
Fixed with #1451.
